common-lib/src/cmr/common/api/web_server.clj

```diff
@@ -46,6 +47,44 @@
   report in Cubby can be in the 5 to 10 MB range."
  (* 50 ONE_MB))
 
+(def buffer-size
+  "Number of bytes to allocate for the buffer used for verifying request body size is not too
+  large. We want to use a smaller buffer size than the max request body size so for small requests
+  we use a small amount of memory to perform the verification."
+  512)
+
+(defn- read-into-buffer
+  "Reads from a stream into the provided byte-array buffer. Returns a map with details about the
+  read operation. The keys in the map are:
+  :at-end? - indicates whether the entire stream has been read.
+  :bytes-read - number of bytes read from the stream. If nothing was read return 0 rather than -1.
+  :buffer-full? - indicates whether the max-bytes-to-read were read."
+  [input-stream buffer offset max-bytes-to-read]
+  (let [bytes-read (.read input-stream buffer offset max-bytes-to-read)]
+    {:at-end? (= -1 bytes-read)
+     :bytes-read (max 0 bytes-read) ;; Java input stream read will return -1 when no bytes are read
+                                    ;; but we want to return the actual bytes read of 0.
+     :buffer-full? (= bytes-read max-bytes-to-read)}))
+
+(def request-body-too-large-response
+  "Response to return when the request body is larger than our allowed MAX_REQUEST_BODY_SIZE."
+  {:status 413
+   :content-type :text
+   :body "Request body exceeds maximum size"})
+
+(defn- byte-stream-from-buffers
+  "Reconstructs request body into a new input stream from all of the buffers used to validate the
+  request body size. This is required so that the request body can be read successfully from
+  within application routes code."
+  [buffers total-bytes]
+  (let [request-body-bytes (->> buffers
+                                persistent!
+                                (reducers/mapcat vec)
+                                reducers/foldcat
+                                (into [])
```

Comment:
oh, wow -- very interesting. Thanks for pasting the error (now I'm super-curious ... my "dig" instincts are twitching ...)